File: sdk/go/engauge.go
package hypermesh

import "context"

// EngaugeApi provides access to Engauge endpoints.
type EngaugeApi struct {
	http *httpClient
}

// Capacity returns capacity metrics.
func (a *EngaugeApi) Capacity(ctx context.Context) (*EngaugeCapacityMetrics, error) {
	var result EngaugeCapacityMetrics
	if err := a.http.get(ctx, "/api/v1/engauge/capacity", &result); err != nil {
		return nil, err
	}
	return &result, nil
}

// Traffic returns traffic metrics.
func (a *EngaugeApi) Traffic(ctx context.Context) (*EngaugeTrafficMetrics, error) {
	var result EngaugeTrafficMetrics
	if err := a.http.get(ctx, "/api/v1/engauge/traffic", &result); err != nil {
		return nil, err
	}
	return &result, nil
}

// MarketplaceListings returns marketplace listings.
func (a *EngaugeApi) MarketplaceListings(ctx context.Context) (*EngaugeListingList, error) {
	var result EngaugeListingList
	if err := a.http.get(ctx, "/api/v1/engauge/marketplace/listings", &result); err != nil {
		return nil, err
	}
	return &result, nil
}

// NodeMetrics returns node metrics.
func (a *EngaugeApi) NodeMetrics(ctx context.Context) (*EngaugeNodeMetrics, error) {
	var result EngaugeNodeMetrics
	if err := a.http.get(ctx, "/api/v1/engauge/node/metrics", &result); err != nil {
		return nil, err
	}
	return &result, nil
}

// Leases returns active leases.
func (a *EngaugeApi) Leases(ctx context.Context) (*EngaugeLeaseList, error) {
	var result EngaugeLeaseList
	if err := a.http.get(ctx, "/api/v1/engauge/leases", &result); err != nil {
		return nil, err
	}
	return &result, nil
}

File: sdk/python/hypermesh/api/engauge.py
"""Engauge API: capacity, traffic, marketplace, node metrics, leases."""

from __future__ import annotations

from typing import Any

from ..types import (
    EngaugeCapacityMetrics,
    EngaugeLeaseList,
    EngaugeListingList,
    EngaugeNodeMetrics,
    EngaugeTrafficMetrics,
)

_PREFIX = "/api/v1/engauge"


class EngaugeApi:
    """Wraps /api/v1/engauge/* endpoints."""

    def __init__(self, transport: Any) -> None:
        self._t = transport

    def capacity(self) -> EngaugeCapacityMetrics:
        data = self._t.get(f"{_PREFIX}/capacity")
        return _parse_capacity(data)

    def traffic(self) -> EngaugeTrafficMetrics:
        data = self._t.get(f"{_PREFIX}/traffic")
        return EngaugeTrafficMetrics(
            organic_ratio=data.get("organic_ratio", 0.0),
            speculative_ratio=data.get("speculative_ratio", 0.0),
            total_requests=data.get("total_requests", 0),
        )

    def marketplace_listings(self) -> EngaugeListingList:
        data = self._t.get(f"{_PREFIX}/marketplace/listings")
        return _parse_listing_list(data)

    def node_metrics(self) -> EngaugeNodeMetrics:
        data = self._t.get(f"{_PREFIX}/node/metrics")
        return EngaugeNodeMetrics(
            activity_score=data.get("activity_score", 0.0),
            receipts=data.get("receipts", 0),
            bandwidth=data.get("bandwidth", 0.0),
        )

    def leases(self) -> EngaugeLeaseList:
        data = self._t.get(f"{_PREFIX}/leases")
        return _parse_lease_list(data)


class AsyncEngaugeApi:
    """Async variant of EngaugeApi."""

    def __init__(self, transport: Any) -> None:
        self._t = transport

    async def capacity(self) -> EngaugeCapacityMetrics:
        data = await self._t.get(f"{_PREFIX}/capacity")
        return _parse_capacity(data)

    async def traffic(self) -> EngaugeTrafficMetrics:
        data = await self._t.get(f"{_PREFIX}/traffic")
        return EngaugeTrafficMetrics(
            organic_ratio=data.get("organic_ratio", 0.0),
            speculative_ratio=data.get("speculative_ratio", 0.0),
            total_requests=data.get("total_requests", 0),
        )

    async def marketplace_listings(self) -> EngaugeListingList:
        data = await self._t.get(f"{_PREFIX}/marketplace/listings")
        return _parse_listing_list(data)

    async def node_metrics(self) -> EngaugeNodeMetrics:
        data = await self._t.get(f"{_PREFIX}/node/metrics")
        return EngaugeNodeMetrics(
            activity_score=data.get("activity_score", 0.0),
            receipts=data.get("receipts", 0),
            bandwidth=data.get("bandwidth", 0.0),
        )

    async def leases(self) -> EngaugeLeaseList:
        data = await self._t.get(f"{_PREFIX}/leases")
        return _parse_lease_list(data)


def _parse_capacity(data: dict[str, Any]) -> EngaugeCapacityMetrics:
    return EngaugeCapacityMetrics(
        bytes_served=data.get("bytes_served", 0),
        compute_delivered=data.get("compute_delivered", 0.0),
        storage=data.get("storage", 0),
        bandwidth=data.get("bandwidth", 0.0),
        uptime=data.get("uptime", 0.0),
    )


def _parse_listing_list(data: dict[str, Any]) -> EngaugeListingList:
    listings = data.get("listings", [])
    return EngaugeListingList(
        count=data.get("count", len(listings)),
        listings=listings,
    )


def _parse_lease_list(data: dict[str, Any]) -> EngaugeLeaseList:
    leases = data.get("leases", [])
    return EngaugeLeaseList(
        count=data.get("count", len(leases)),
        leases=leases,
    )
